### PR TITLE
Split config unmarshaler per config section

### PR DESCRIPTION
**Pull request by bogdandrutu**
This change will allow to break circular dependency (service | service/internal/pipelines | service/internal/configunmarshaler | config.Config) which blocks removing of the deprecated `config.Config`

### CODE REVIEW

File: service/internal/configunmarshaler/defaultunmarshaler.go [hunk 1]
```diff
@@ -51,41 +47,25 @@ type configError struct {
...
-	// receiversKeyName is the configuration key name for receivers section.
-	receiversKeyName = "receivers"
-
-	// exportersKeyName is the configuration key name for exporters section.
-	exportersKeyName = "exporters"
-
-	// processorsKeyName is the configuration key name for processors section.
-	processorsKeyName = "processors"
-)
+const ()
```
Comment (codeboten) [inline service/internal/configunmarshaler/defaultunmarshaler.go:50]: Is this still needed?

File: confmap/confmap.go [hunk 2]
```diff
@@ -273,20 +273,25 @@ func unmarshalerHookFunc(result interface{}) mapstructure.DecodeHookFuncValue {
 		}
 
 		toPtr := to.Addr().Interface()
-		if _, ok := toPtr.(Unmarshaler); !ok {
+		// Need to ignore the top structure to avoid circular dependency.
+		if toPtr == result {
```
Comment (codeboten) [inline confmap/confmap.go:277]: Is this moved up to avoid an unnecessary allocation for `unmarshaler`?

**Review by codeboten — APPROVED** [review]

File: service/internal/configunmarshaler/defaultunmarshaler.go (hunk 1 above)
Comment (bogdandrutu) [reply service/internal/configunmarshaler/defaultunmarshaler.go:50]: Nope. Removed.

File: confmap/confmap.go (hunk 2 above)
Comment (bogdandrutu) [reply confmap/confmap.go:277]: You caught me :)) Now, I am blocked by https://github.com/open-telemetry/opentelemetry-collector/pull/6392

**Pull request merged**